### PR TITLE
feat: add safe cast u256 <> u64 for bridge operations

### LINKED ISSUE #38: [Feature]: Safe u256 <> u64 conversion
### What is the feature you would like to see?

Add a function to convert `u256` ETH balance into `u64` and vice versa. 

DeFi projects on Sui use `u64` as the main data type once Ethereum uses `u256`. It brings potential security issues with brigdes and other use cases.

### PR DESCRIPTION
**Pull request by bidzyyys**
Resolves #38 

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [ ] Changelog

## Summary by CodeRabbit

* **New Features**
  * Added decimal-scaling utilities for safe token-amount conversions with input validation, truncation semantics when downscaling, and overflow protection.
  * Supports reliable upcasting/downcasting between different token decimal representations.

* **Tests**
  * Added comprehensive tests covering identical and varying decimals, edge cases, truncation behavior, overflow/error conditions, and round-trip conversion integrity.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by bidzyyys** [comment]
> Hey @bidzyyys,
> 
> I don’t think a generic `from_u256` is what Damir had in mind for this issue. I believe he was thinking about a function that **specifically handles downcasting balances into `u64`**, which is the balance type used in `coin::Coin`. We can of course follow up with him to confirm.
> 
> This function should deal with **decimal scaling and truncation**, since that’s exactly what bridges need to do. A `u256` balance can hold up to 77 digits, while a `u64` can hold only 20, so representing a token with 18 decimals using Sui’s `Coin<T>` is not feasible without a proper downcast.
> 
> We should add something like:
> 
> ```move
> downcast_balance(raw_amount: u256, raw_decimals: u8, scaled_decimals: u8): u64
> ```
> 
> The return value would be the scaled balance, potentially truncated, while preserving the economic meaning of the amount.
> 
> Note that it doesn't make much sense to implement this function for return types different than u64 since it is specifically designed for coins,

Based on Damir's response, we should take a look into Axelar's implementation.

File: math/core/sources/coin_utils.move [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+public fun safe_downcast_balance(raw_amount: u256, source_decimals: u8, target_decimals: u8): u64 {
+    // Validate decimal ranges.
+    validate_decimals(source_decimals, target_decimals);
+
+    let scaled_amount = scale_amount(raw_amount, source_decimals, target_decimals);
+
+    // Verify it fits in `u64`.
+    assert!(scaled_amount <= (std::u64::max_value!() as u256), ESafeDowncastOverflowedInt);
+
+    (scaled_amount as u64)
```
Comment (immrsd) [inline math/core/sources/coin_utils.move]: It doesn't require brackets here
```suggestion
    scaled_amount as u64
```

File: math/core/tests/coin_utils_tests.move [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+    // 1 token (9 decimals) -> 0 decimals (truncates fractional part)
+    let amount = 1000000000u256;
+    let result = coin_utils::safe_downcast_balance(amount, 9, 0);
+    assert_eq!(result, 1);
+}
+
+#[test]
+fun test_downcast_max_decimal_diff() {
+    // Test maximum decimal difference (24 -> 0)
+    let amount = 1000000000000000000000000u256; // 1 token with 24 decimals
```
[the remark this answers is not in the document]
Comment (bidzyyys) [reply math/core/tests/coin_utils_tests.move]: Good point!

**Review by immrsd — APPROVED** [review]
LGTM
